File: apps/desktop/trees/trees.scala
package com.fishuyo.seer
package examples.trees
import maths._
import spatial._
import graphics._
import trees._
// import particle.structures._
import io._
import io.kinect._
import dynamic._
import audio._

import util._

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.math.Matrix4
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.GL20

import collection.mutable.ListBuffer


object Main extends SeerApp with AudioSource {

  // DesktopApp.loadLibs()
  // Scene.push(this)
  //Scene.push(Kinect)
  
  var ground:Model = _
  var tree = new Tree() 
  tree.branch()

  var wind:com.badlogic.gdx.audio.Music = _
  Audio.push(this) 

  // var niceView = new Pose(Vec3(0.7285795f,0.f, 1.4537674f), Quat(0.9765499f, 0.113466285f, 0.18053834f, -0.029700035f))
  var niceViews = new ListBuffer[Pose]()
  var niceTrees = new ListBuffer[Array[Float]]()
  niceViews += new Pose(Vec3(0.7285795f,0.f, 1.4537674f), Quat(0.9765499f, 0.113466285f, 0.18053834f, -0.029700035f))

  var rms = 0.f

  var volume = Vec3(0)
  var color = Vec3(68.0/255.0,122.0/256.0,222.0/256.0)

  var day = Vec3(1.f,0,1.f)
  var useLocalZ = Vec3(1.f)
  var dayUniforms = Vec3(0.f,4.1f,0.f)
  var nightUniforms = Vec3(0.f,2.f,1.f)

  var move = Vec3(0.f)
  var rot = Vec3(0.f)
  var nmove = Vec3(0.f)
  var nrot = Vec3(0.f)

  var pref = ""

  var live:Ruby = null

  var theta = 0.f
  var dw = 0.f
  var blurDist = 0.1f

  // DesktopApp() 

  override def init(){
    ground = OBJ(pref + "res/landscapealien.obj") //new ObjLoader().loadObj(Gdx.files.internal("src/main/scala/drone/landscapealien.obj"))
    ground.pose.quat.set(0.42112392f,-0.09659095f, 0.18010217f, -0.8836787f)
    ground.pose.pos.set(0.f,-1.3f,-.0f)
    ground.scale.set(5.f,5.f,5.f)

    val f = Gdx.files.internal(pref +"res/wind.mp3")
    wind = Gdx.audio.newMusic(f)
    wind.setVolume(0.0f)
    wind.setLooping(true)
    wind.play

    var s = Shader.load("firstPass", Gdx.files.internal(pref+"res/shaders/firstPass.vert"), Gdx.files.internal(pref+"res/shaders/firstPass.frag"))
    s.monitor()
    s = Shader.load("secondPass", Gdx.files.internal(pref+"res/shaders/secondPass.vert"), Gdx.files.internal(pref+"res/shaders/secondPass.frag"))
    s.monitor()
    s = Shader.load("test", Gdx.files.internal(pref+"res/shaders/test.vert"), Gdx.files.internal(pref+"res/shaders/test.frag"))
    s.monitor()
    s = Shader.load("texture", Gdx.files.internal(pref+"res/shaders/texture.vert"), Gdx.files.internal(pref+"res/shaders/texture.frag"))
    s.monitor()
    s = Shader.load("composite", Gdx.files.internal(pref+"res/shaders/composite.vert"), Gdx.files.internal(pref+"res/shaders/composite.frag"))
    s.monitor()

    SceneGraph.root.shader = "firstPass"

    println("trees init.")
    val node = new RenderNode
    node.shader = "secondPass"
    val quad = new Drawable {
      val m = Mesh(Primitive2D.quad)
      override def draw(){
        if( day.x == 1.f) Shader("secondPass").setUniformf("u_depth", 0.f)
        else Shader("secondPass").setUniformf("u_depth", 1.f)

        m.draw()
      }
    }
    node.scene.push( quad )
    SceneGraph.root.outputTo(node)

    val node2 = new RenderNode
    node2.shader = "test"
    val quad2 = new Drawable {
      val m = Mesh(Primitive2D.quad)
      override def draw(){
        Shader("test").setUniformf("u_dist", getBlurDist)
        m.draw()
      }
    }
    node2.scene.push( quad2 )
    // node.outputTo(node2)

    val compNode = new RenderNode
    compNode.shader = "composite"
    compNode.clear = false
    val quag = new Drawable {
      val m = Mesh(Primitive2D.quad)
      override def draw(){
        // Shader("composite").setUniformf("u_blend0", 1.0f)
        // Shader("composite").setUniformf("u_blend1", 1.0f)
        // Shader("composite").setUniformMatrix("u_projectionViewMatrix", new Matrix4())
        m.draw()
      }
    }
    compNode.scene.push( quag )
    node.outputTo(compNode)
    compNode.outputTo(compNode)
    compNode.outputTo(ScreenNode)

    // Kinect.init()
    tree.init()
    live = new Ruby(pref+"trees.rb")
  }
  override def draw(){

    theta += dw
    MatrixStack.rotate(0.f,theta,0.f)

    if(day.x == 1.f){
      Shader("firstPass").setUniformf("u_useLocalZ", dayUniforms.z)
      Shader("firstPass").setUniformf("u_near", dayUniforms.x)
      Shader("firstPass").setUniformf("u_far", dayUniforms.y)
      Shader("firstPass").setUniformf("u_useTexture", 0.f)
      ground.draw()
      Shader("firstPass").setUniformf("u_useLocalZ", dayUniforms.z)
      Shader("firstPass").setUniformf("u_near", dayUniforms.x)
      Shader("firstPass").setUniformf("u_far", dayUniforms.y)
      Shader("firstPass").setUniformf("u_useTexture", 1.f)
      // if(day.z == 1.f) Kinect.draw()
      tree.draw()
    }else{
      Shader("firstPass").setUniformf("u_useLocalZ", 0.f)
      Shader("firstPass").setUniformf("u_near", 0.f)
      Shader("firstPass").setUniformf("u_far", 4.1f)
      Shader("firstPass").setUniformf("u_useTexture", 0.f)
      ground.draw()
      Shader("firstPass").setUniformf("u_useLocalZ", nightUniforms.z)
      Shader("firstPass").setUniformf("u_near", nightUniforms.x)
      Shader("firstPass").setUniformf("u_far", nightUniforms.y)
      Shader("firstPass").setUniformf("u_useTexture", 1.f)
      //Kinect.draw()
      tree.draw()
    }

  }

  def rotateWorld(v:Float) = dw = v
  def blurDist(v:Float){ blurDist = v }
  def getBlurDist() = blurDist
  
  def draw2(){
      // Shader("secondPass").setUniformf("u_edge", 1.)
    if( day.x == 1.f) Shader("secondPass").setUniformf("u_depth", 0.f)
    else Shader("secondPass").setUniformf("u_depth", 1.f)
  }
  override def animate(dt:Float){

    live.animate(dt)
    // Kinect.animate(dt)
    // MatrixStack.worldPose.rotate(0,dw,0)
    tree.animate(dt)
  }

  override def audioIO( in:Array[Float], out:Array[Array[Float]], numOut:Int, numSamples:Int){
    var sum = 0.f
    for( i<-(0 until numSamples)){ //Audio.bufferSize)){
      sum += in(i)*in(i) //Audio.in(i)*Audio.in(i)
      rms = math.sqrt( sum / numSamples ).toFloat
    }
    //println( rms )
  }
}

File: apps/desktop/trees/trees.rb
# require 'java'

module M
  include_package "com.fishuyo.seer"
  include_package "com.fishuyo.seer.io"
  include_package "com.fishuyo.seer.io.kinect"
  include_package "com.fishuyo.seer.maths"
  include_package "com.fishuyo.seer.spatial"
  include_package "com.fishuyo.seer.graphics"
  include_package "com.fishuyo.seer.util"
  include_package "com.fishuyo.seer.examples.trees"
  include_package "com.fishuyo.seer.trees"
end

class Object
  class << self
    alias :const_missing_old :const_missing
    def const_missing c
      M.const_get c
    end
  end
end

$Seer = Java::com.fishuyo.seer
# Seer = $Seer
$Vec3 = $Seer.maths.Vec3
$Camera = $Seer.graphics.Camera
$Audio = $Seer.audio.Audio
$OSC = $Seer.io.OSC
$Pad = $Seer.io.Trackpad

$tree = $Seer.examples.trees.Main.tree
tree = $tree

# $tree.root.position.set(0,-0.5,0)
# $tree.root.pin($tree.root.pose.pos)



def tree0
	$tree.branchNum.setChoices([0,1,2].to_java :int)
	$tree.branchNum.setProb([0.0,1.0,0.0].to_java :float)
	$tree.sLength.setMinMax(0.5,0.5,false)

	$tree.bAngle.z.setMinMax(0.0,0.0,false)
	$tree.bAngle.y.setMinMax(0.0,1.2,true)
	$tree.bAngle.x.setMinMax(0.0,0.0,false)

	$tree.sAngle.z.setMinMax(1.0,1.0,false)
	$tree.sAngle.y.setMinMax(0.0,0.0,false)
	$tree.sAngle.x.setMinMax(0.0,0.0,false)
end


depth = 10

tree0()

$tree.setAnimate(true)
$tree.setReseed(true)
$tree.setDepth(depth)
$tree.branch(depth)


#$Seer.trees.Fabric.gv.set(0.0,-5.0,0.0)
$Seer.trees.Trees.gv.set(0.0,1.0,0.0)


### Trackpad state ###
mx=0.0
my=1.0
mz=1.0 
rz=0.1
rx=0.0
ry=0.0
$Pad.clear()
$Pad.connect()
$Pad.bind( lambda{|i,f|
	t = Main.tree #Seer.examples.trees.Main.tree
	if i == 1
		ur = SceneGraph.root.camera.nav.ur()
		uf = SceneGraph.root.camera.nav.uf()

		# Main.tree.root.applyForce( (v * 10.0 * Main.rms) )
		# t.root.applyForce( Vec3.apply((f[0]-0.5)*1.0*f[4],0,(f[1]-0.5)*f[4]))
		t.root.applyForce( ur*(f[0]-0.5)*2.0*f[4])
		t.root.applyForce( uf*(f[1]-0.5)*-2.0*f[4])

		Shader.apply("composite")
		Shader.shader.get.uniforms.update("u_blend0", f[0])
		Shader.shader.get.uniforms.update("u_blend1", f[1])

	elsif i == 2
		mx = mx + f[2]*0.05  
		mz = mz + f[3]*-0.05
	elsif i == 3
		mx = mx + f[2]*0.05  
		my = my + f[3]*0.05
	elsif i == 4
		rz = rz + f[3]*0.05
		rx = rx + f[2]*0.05
	end

	if i > 1
		t.bAngle.y.setMinMax( 0.05, mx,false )
		#t.bAngle.y.set(mx)
	    t.sRatio.setMinMax( 0.05, mz, false )
	    #t.sRatio.set( mz )
	    t.bRatio.setMinMax( 0.05, my, false )
	    #t.bRatio.set( my )
	    t.sAngle.x.setMinMax( 0.05, rx, false )
	    t.bAngle.x.setMinMax( 0.05, rx, false )
	    #t.sAngle.x.set( rx )
	    t.sAngle.z.setMinMax( 0.05, rz, false )
	    t.bAngle.z.setMinMax( 0.05, rz, false )
	    #t.sAngle.z.set( rz )
	    #t.branch(depth)
	    t.refresh()

	    # t.root.accel.zero
	    # t.root.euler.zero
	end
})


move = Vec3.apply(0,0,0)
rot = Vec3.apply(0,0,0)
nmove = Vec3.apply(0,0,0)
nrot = Vec3.apply(0,0,0)

# Kinect.connect()
# Kinect.startDepth()
# Kinect.blob.clear()
Kinect.bind( lambda{|r|
	#if f[2] < 10 || f[3] < 10 then return end

	if r.size == 0
		return
	end

	t = Main.tree
	size = r[0].w*r[0].h/10000.0
	x = (r[0].x - 320)/320.0
	y = (r[0].y - 240)/-240.0
	a = r[0].angle
	if a < 90.0
		a = -a
	else
		a = 180.0 - a
	end

	a = a / 180.0
	i = 0
	# puts a
	# print i
	# print " "
	# print x
	# print " "
	# print y
	# print " "	
	# # print f[2]
	# # print " "
	# # puts f[3]
	# puts size
		


	if Main.day.x == 1.0
		ur = SceneGraph.root.camera.nav.ur()
		uf = SceneGraph.root.camera.nav.uf()
		t.root.applyForce( ur*(-x*20))
		t.root.applyForce( uf*(y*20))
		Main.volume.lerpTo(Vec3.apply(1,1,1),0.01)

		return
	end

	# puts i
	nm = Main.nmove
	nr = Main.nrot
	if i == 0
		Main.nrot.set(0.0,0.0,0.0)
		Main.nmove.set(a*2,nm.y,nm.z)
		# Main.nmove.set(nm.x, nm.y, (y+0.5)*size+1.0 )
		# Main.nrot.set(x*2.0, nr.y, nr.z)
	elsif i == 1
		Main.nmove.set(x*4.0, (y+0.5)*size+0.8, nm.z )
	elsif i == 2
		Main.nrot.set(nr.x,nr.y, x*4.0)
	elsif i == 4
		rz = f[3]*0.05
		rx = f[2]*0.05
	end

	# move.lerpTo(nmove, 0.05 )
	
	# t.bAngle.y.setMinMax( 0.05, Main.move.x,false )
	# #t.bAngle.y.set(mx)
 #    t.sRatio.setMinMax( 0.05, Main.move.z, false )
 #    #t.sRatio.set( mz )
 #    t.bRatio.setMinMax( 0.05, my, false )
 #    #t.bRatio.set( my )
 #    t.sAngle.x.setMinMax( 0.05, rx, false )
 #    t.bAngle.x.setMinMax( 0.05, rx, false )
 #    #t.sAngle.x.set( rx )
 #    t.sAngle.z.setMinMax( 0.05, rz, false )
 #    t.bAngle.z.setMinMax( 0.05, rz, false )
 #    #t.sAngle.z.set( rz )
 #    #t.branch(depth)
 #    t.refresh()

 #    t.root.accel.zero
 #    t.root.euler.zero
	
})

Trees.setDamp(150.0)

# Kinect.threshold.set( 0.0, 0.4, 0.0 )
# Kinect.device.get().setTiltAngle(0.0)
# Kinect.bgsub.updateBackgroundNextFrame()
# Kinect.bgsub.setThreshold(10.0)

# Kinect.setSizeThreshold( 20 )

def animate(dt)

	# f = com.fishuyo.seer.util.Random.float.apply()
	# # Main.blurDist(f*0.001 + 0.001)
	# Main.blurDist(0)

	t = Main.tree
	# t.root.pose.pos.set(0,-0.5,0)

	if DesktopApp.app.frameCount % 10
		Main.wind.setVolume(Main.volume.x)
	end
	Main.volume.lerpTo(Vec3.apply(0.0,0,0),0.01)

	dawnf = 200
	dayf = 3250
	duskf = 3700
	nightf = 6500

	blue = Vec3.apply(68.0/255.0,122.0/256.0,222.0/256.0)
	white = Vec3.apply(0.1)

	gpose = Pose.apply(Vec3.apply(0,-1.3,0),Quat.apply(0.42112392,-0.09659095, 0.18010217, -0.8836787))

	frame = DesktopApp.app.frameCount % (6500)
	#puts frame
	if frame <= dawnf
		Main.nmove.set(1.3, 1.22, 1.2)
		Main.nrot.set(0.6, 0, 1.6)
		# Main.ground.pose.quat.slerpTo(gpose.quat, 1/250.0)
  #   	Main.ground.pose.pos.lerpTo(gpose.pos,1/250.0)
  #   	Main.ground.scale.lerpTo(Vec3.apply(10.0,10.0,10.0),1/50.0)

	elsif frame <= dayf
		Main.color.lerpTo(blue, 0.01)
		#Main.ground.pose.set(gpose)
		#Main.ground.scale.set(10.0,10.0,10.0)
		Main.day.set(1,0,1)

	elsif frame <= duskf
		if frame <= duskf-100
			Main.color.lerpTo(Vec3.apply(0.7,0,0.3), 0.01)
		else
			Main.color.lerpTo(white, 0.01)
		end
    	# Main.ground.scale.lerpTo(Vec3.apply(-1.0,-1.0,-1.0),1/200.0)

	elsif frame <= nightf
		Main.color.set(white)
		Main.day.set(0,0,0)
		Main.nightUniforms.set(-1,2,1)

	end

	# Main.day.set(0,0,0)

	Shader.setBgColor(RGBA.apply(Main.color,1.0))

	if Main.day.x == false #0.0
		Main.move.lerpTo(Main.nmove, 0.05 )
		Main.rot.lerpTo(Main.nrot, 0.05 )

		t.bAngle.y.setMinMax( 0.0, Main.move.x,false )
		#t.bAngle.y.set(mx)
	    t.sRatio.setMinMax( 0.05, Main.move.z, false )
	    #t.sRatio.set( mz )
	    t.bRatio.setMinMax( 0.05, Main.move.y, false )
	    #t.bRatio.set( my )
	    t.sAngle.x.setMinMax( 0.0, Main.rot.x, false )
	    t.bAngle.x.setMinMax( 0.0, Main.rot.x, false )
	    #t.sAngle.x.set( rx )
	    t.sAngle.z.setMinMax( 0.0, Main.rot.z, false )
	    t.bAngle.z.setMinMax( 0.0, Main.rot.z, false )
	    #t.sAngle.z.set( rz )
	    #t.branch(depth)
	    t.refresh()

	    # t.root.accel.zero
	    # t.root.euler.zero
	end

	height = t.root.getMaxHeight() + 1
	height = 10 if height > 10
	height = 1 if height < 1
	ncamPos = SceneGraph.root.camera.nav.uf()*-height
	SceneGraph.root.camera.nav.pos.lerpTo( ncamPos, 0.005 )
	Main.dayUniforms.set(0.0,10.0,0.0)

	pos = SceneGraph.root.camera.nav.pos + SceneGraph.root.camera.nav.uf()*1.5
	pos += SceneGraph.root.camera.nav.ur()*(0.8)
	pos += SceneGraph.root.camera.nav.uu()*-0.5
	Kinect.cube.pose.pos.lerpTo( pos, 0.1)
	Kinect.cube.pose.quat.slerpTo( SceneGraph.root.camera.nav.quat, 0.1)
	scl = 0.5
	Kinect.cube.scale.set( scl*1.0, scl*480.0/640.0, 0.05 )
	#puts Kinect.cube.pose.pos
	#puts Kinect.cube.pose.quat

	Kinect.cube.pose.pos.set(0.49894238, -0.6203693, 1.2422245)
	Kinect.cube.pose.quat.set(-0.7918381, 0.60365784, -0.08232442, -0.042571533)


	puts ScreenNode.inputs.apply(0).camera.nav.pos


end

Keyboard.clear()
Keyboard.use()
Keyboard.bind("r", lambda{
	Main.rotateWorld(0.004)	
})
Keyboard.bind("t", lambda{
	Main.rotateWorld(0.000)	
})
Keyboard.bind("n", lambda{
	p = Pose.apply(Camera.nav)
	Main.niceViews += p
	Main.niceTrees += [mx,my,mz,rx,ry,rz].to_java 

})
Keyboard.bind("m", lambda{
	for i in 0..Main.niceTrees.length
		puts Main.niceTrees(i)
	end
})
